// @hasura/ndc-lambda-sdk/dist/src/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.unreachable = void 0;
exports.isArray = isArray;
exports.throwError = throwError;
exports.mapObjectValues = mapObjectValues;
exports.getFlags = getFlags;
exports.sleep = sleep;
const unreachable = (x) => { throw new Error(`Unreachable code reached! The types lied! 😭 Unexpected value: ${x}`); };
exports.unreachable = unreachable;
function isArray(x) {
    return Array.isArray(x);
}
// Throws an error. Useful for using after a short-circuiting ?? operator to eliminate null/undefined from the type
function throwError(...args) {
    throw new Error(...args);
}
function mapObjectValues(obj, fn) {
    return Object.fromEntries(Object.entries(obj).map(([prop, val]) => [prop, fn(val, prop)]));
}
/**
 * Returns all the set bitwise flags in a value, where the flags are defined on an enum type.
 * Useful for debugging TypeScript API types (eg ts.Type.flags with enum ts.TypeFlags)
 */
function getFlags(flagsEnum, value) {
    return Object
        .keys(flagsEnum)
        .flatMap(k => {
        const k_int = parseInt(k);
        return isNaN(k_int)
            ? []
            : (value & k_int) !== 0
                ? [flagsEnum[k]]
                : [];
    });
}
function sleep(ms) {
    return new Promise((r) => setTimeout(r, ms));
}
